gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/sql/execute/MemberLogsMessage.java

```diff
@@ -0,0 +1,392 @@
+        stringBuffer.append(getOffsetBytes(
+            files[i].getAbsolutePath(), fileLengths[i], effectiveStartIndex, effectiveEndIndex));
+      }
+      sum += fileLengths[i];
+    }
+
+    return stringBuffer.toString();
+  }
+
+  /** Return a string containing part of a file from byte 'start' to 'end'. */
+  private String getOffsetBytes(String filePath, Long fileLength, Long start, Long end){
+
+    File file = new File(filePath);
+    if(file == null || StringUtils.isEmpty(filePath)){
+      logger.debug("File is not specified.." + filePath);
+      return "File is not specified..";
+    }
+
+    Long effectiveStart = Math.max(0, start);
+    Long effectiveEnd = Math.min(fileLength, end);
+
+    int buffLength = (int)(effectiveEnd-effectiveStart);
+    byte[] buff = new byte[buffLength];
+    FileInputStream fis = null;
+    try {
+      fis = new FileInputStream(file);
+      if(fis != null){
+        ByteStreams.skipFully(fis, effectiveStart);
+        ByteStreams.readFully(fis, buff);
+      } else {
+        logger.debug("Could not read File [ " + filePath + " ]");
+        return "Could not read File [ " + filePath + " ]";
+      }
+
+    } catch (FileNotFoundException e) {
+      logger.error("FileNotFoundException: " + e.getMessage(), e);
+    } catch (IOException e) {
+      logger.error("IOException: "+e.getMessage(), e);
+    } finally {
+      if(fis != null){
+        try {
+          fis.close();
+        } catch (IOException e) {
+          e.printStackTrace();
+        }
+      }
+    }
+
+    return new String(buff);
+  }
+
+  @Override
+  public Set<DistributedMember> getMembers() {
+    Set<DistributedMember> members = new HashSet<DistributedMember>();
+    if(this.memberId != null || !this.memberId.isEmpty()) {
+      Set<DistributedMember> distMembers = getAllGfxdMembers();
+      Iterator<DistributedMember> itr = distMembers.iterator();
+      while(itr.hasNext()){
+        DistributedMember distMember = itr.next();
+        if(distMember.getId().equalsIgnoreCase(memberId)){
+          members.add(distMember);
+          break;
+        }
+      }
+    }
+    return members;
+  }
+
+  @Override
+  public void fromData(DataInput in)
+      throws IOException, ClassNotFoundException {
+    super.fromData(in);
+    this.memberId = DataSerializer.readString(in);
+    this.offset = DataSerializer.readObject(in);
+    this.byteLength = DataSerializer.readObject(in);
+  }
+
+  @Override
+   public void toData(final DataOutput out) throws IOException {
+    super.toData(out);
+    DataSerializer.writeString(this.memberId, out);
+    DataSerializer.writeObject(this.offset , out);
+    DataSerializer.writeObject(byteLength, out);
+  }
+
+  @Override
+  public void postExecutionCallback() {
+
+  }
+
+  @Override
+  public boolean isHA() {
+    return false;
+  }
+
+  @Override
+  public boolean optimizeForWrite() {
+    return false;
+  }
+
+  @Override
+  protected GfxdFunctionMessage clone() {
+    return null;
```

Comment:
This can cause trouble. Provide a simple shallow clone or call a copy constructor which assigns the fields.